File: WaveMobileChallenge/MainViewController.m
//
//  MainViewController.m
//  WaveMobileChallenge
//
//  Created by Leon Chen on 2016-03-24.
//  Copyright © 2016 LeonChen. All rights reserved.
//

#import "MainViewController.h"
#import "ProductFetcherService.h"
#import "Product.h"

@interface MainViewController () <UITableViewDataSource>

@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicator;

@property (strong, nonatomic) NSArray<Product *> *products;

@property (nonatomic, assign) BOOL loading;

@end

@implementation MainViewController

#pragma mark - View Controller Life Cycle

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    
    self.title = @"Products";
    
    self.tableView.dataSource = self;
}

-(void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    if (!self.products)
    {
        [self loadData];
    }
}

#pragma mark - Other Methods

- (void) setLoading:(BOOL)loading
{
    _loading = loading;
    
    if (_loading)
    {
        [self.activityIndicator startAnimating];
        [self.tableView setHidden:YES];
    }
    else
    {
        [self.activityIndicator stopAnimating];
        [self.tableView setHidden:NO];
    }
}

- (void)loadData
{
    self.loading = YES;
    
    __weak typeof(self) weakSelf = self;
    
    [self.productFetcherService fetchProducts:^(NSArray<Product *> *products) {
        
        weakSelf.loading = NO;
        
        weakSelf.products = products;
        [weakSelf.tableView reloadData];
        
    } failure:^(NSError *error) {
        
        weakSelf.loading = NO;
        [weakSelf showRetryDialog];
        
    }];
}

- (void)showRetryDialog
{
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Error" message:@"Couldn't load products data!" preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction *retryAction = [UIAlertAction actionWithTitle:@"Retry" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action)
    {
        [self loadData];
    }];
    
    [alertController addAction:retryAction];
    
    [self presentViewController: alertController animated:YES completion:nil];
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.products.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *cellId = @"TableCell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier: cellId];
    
    if (cell == nil)
    {
        cell = [[UITableViewCell alloc] initWithStyle: UITableViewCellStyleSubtitle reuseIdentifier: cellId];
    }
    
    Product *product = self.products[indexPath.row];
    
    cell.textLabel.text = product.name;
    cell.detailTextLabel.text = [NSString stringWithFormat: @"$%.2f", product.price];
    
    return cell;
}

@end

File: WaveInterView/WaveInterView/ViewControllers/MainViewController.swift
//
//  MainViewController.swift
//  WaveInterView
//
//  Created by Zefeng Qiu on 2016-12-07.
//  Copyright © 2016 Zefeng Qiu. All rights reserved.
//

import UIKit

class MainViewController: BaseViewController {
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    WaveClient.sharedInstance.getProductListBaseOnBusinessID(completionHandlerForGetList: { success, error in
      if success {
        GCD.performUIUpdatesOnMain({
          self.tableView.delegate = self
          self.tableView.dataSource = self
          self.setupTableView()
          WaveProductTableViewCell.registerInTableView(tableView: self.tableView)
        })
      }else {
        if let error = error {
          self.presentAlert(title: "Error", message: error)
        }
      }
    })
    
  }
  
  override func didReceiveMemoryWarning() {
    super.didReceiveMemoryWarning()
    // Dispose of any resources that can be recreated.
  }
  
}

//MARK: UITableView DataSource
extension MainViewController: UITableViewDataSource {
  
  func numberOfSections(in tableView: UITableView) -> Int {
    return 1
  }
  
  func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return WaveClient.sharedInstance.products.count
  }
  
  func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = WaveProductTableViewCell(waveProduct: WaveClient.sharedInstance.products[indexPath.row])
    return cell
  }
  
}


//MARKL: UITableView Delegate
extension MainViewController: UITableViewDelegate {
  

}

